### PR TITLE
Prepare to purge resource API

### PR DESCRIPTION
**Pull request by sequencer**
I split the traditional Resource Binding out from BareSubsystem, I should not break any dependency on `BaseSubsytem`.
But for a modern API in the future, I prepare to land the brand new `Property` API in Chisel and `OM` API in CIRCT for metadata processing.

### CODE REVIEW

File: src/main/scala/subsystem/BaseSubsystem.scala [hunk 1]
```diff
@@ -19,22 +19,29 @@ case object InTile extends HierarchicalLocation("InTile")
...
   lazy val json = JSON(bindingTree)
 }
 
-abstract class BareSubsystemModuleImp[+L <: BareSubsystem](_outer: L) extends LazyRawModuleImp(_outer) {
-  val outer = _outer
-  ElaborationArtefacts.add("graphml", outer.graphML)
-  ElaborationArtefacts.add("dts", outer.dts)
-  ElaborationArtefacts.add("json", outer.json)
+trait HasDtsImp[+L <: HasDts] extends LazyRawModuleImp {
+  def dtsLM: L
```
Comment (jerryz123) [inline src/main/scala/subsystem/BaseSubsystem.scala:31]: What does "LM" stand for?

File: src/main/scala/subsystem/BaseSubsystem.scala [hunk 2]
```diff
@@ -19,22 +19,29 @@ case object InTile extends HierarchicalLocation("InTile")
 case object InSubsystem extends HierarchicalLocation("InSubsystem")
 case object InSystem extends HierarchicalLocation("InSystem")
 
-/** BareSubsystem is the root class for creating a subsystem */
-abstract class BareSubsystem(implicit p: Parameters) extends LazyModule with BindingScope {
+// HasDts is generating metadatas from Scala, which is not the target for new diplomacy and Property.
+// It will be deprecated and removed after we migrate all metadata handling logic to OM Dialect.
+trait HasDts extends LazyModule with BindingScope {
```
Comment (jerryz123) [inline src/main/scala/subsystem/BaseSubsystem.scala:24]: Nit: should be `HasDTS`

Comment (michael-etzkorn) [reply src/main/scala/subsystem/BaseSubsystem.scala:24]: FWIW: camel casing like this is a common style suggestion: 
 https://google.github.io/styleguide/javaguide.html 5.3

File: src/main/scala/subsystem/BaseSubsystem.scala (hunk 1 above)
Comment (sequencer) [reply src/main/scala/subsystem/BaseSubsystem.scala:31]: LazyModule, aka outer. Plz suggest me some better name ;p

File: src/main/scala/subsystem/BaseSubsystem.scala [hunk 3]
```diff
@@ -19,22 +19,29 @@ case object InTile extends HierarchicalLocation("InTile")
...
   lazy val dtb = DTB(dts)
   lazy val json = JSON(bindingTree)
 }
 
-abstract class BareSubsystemModuleImp[+L <: BareSubsystem](_outer: L) extends LazyRawModuleImp(_outer) {
-  val outer = _outer
-  ElaborationArtefacts.add("graphml", outer.graphML)
-  ElaborationArtefacts.add("dts", outer.dts)
-  ElaborationArtefacts.add("json", outer.json)
+trait HasDtsImp[+L <: HasDts] extends LazyRawModuleImp {
```
[the remark this answers is not in the document]
Comment (sequencer) [reply src/main/scala/subsystem/BaseSubsystem.scala:30]: nice catch!

File: src/main/scala/subsystem/BaseSubsystem.scala (hunk 1 above)
Comment (jerryz123) [reply src/main/scala/subsystem/BaseSubsystem.scala:31]: Ah ok, this is fine, I just didn't parse that

File: src/main/scala/subsystem/BaseSubsystem.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (sequencer) [reply src/main/scala/subsystem/BaseSubsystem.scala:30]: Thought about it: it at least should be:
```suggestion
trait HasDtsImp[+L <: HasDts] { this: LazyRawModuleImp =>
```
Is this OK?

[the remark this answers is not in the document]
Comment (jerryz123) [reply src/main/scala/subsystem/BaseSubsystem.scala:30]: LGTM

**Pull request merged**